### PR TITLE
[RHELC-297] Clean up kmods functionality

### PR DESCRIPTION
**Pull request by r0x0d**
These changes are cleanup of most of the things that were left behind from
PR #193.

Some of the functions were changed to make it easier to read and debug if
necessary, some of them were moved around to the files that made sense to place
they and some of them just needed a docstring to tell what the function was all
about.

This PR it is introduced a unit test for the function `get_enabled_rhel_repos`
since that was introduced in the past and never had a unit test tied to it.

Jira reference (If any): https://issues.redhat.com/browse/OAMG-4668
Bugzilla reference (If any):

Signed-off-by: Rodolfo Olivieri <rolivier@redhat.com>

### CODE REVIEW

File: convert2rhel/checks.py
Comment (r0x0d) [inline convert2rhel/checks.py:373]: I was unsure about this... Maybe we should move this function to `convert2rhel/repos.py` file? Since it's dealing with "repos" even it is only extracting the version from a string.

Not sure if the `convert2rhel/repos.py` or `convert2rhel/utils.py` is a better location for this function.

File: convert2rhel/systeminfo.py
Comment (r0x0d) [inline convert2rhel/systeminfo.py:313]: This `TODO` here needs any new information? It's not telling much. 

If it doesn't make sense, can we remove it?

File: convert2rhel/checks.py
Comment (abadger) [reply convert2rhel/checks.py]: It feels like maybe a function for pkghandler?  (Unless we want pkghandler to only deal with packages-in-repositories).  It's parsing a version from a single package's NEVR.  One possible reason not to move this (or to change it when we do move it), though: right now, this function is very specifically targeted at the NEVR of kernel package's that we think we are going to encounter.  All of the following NEVRs would output the same value from this function despite being different packages:

* kernel-2.4.0-1.el8_1
* kernel-2.4-0.1.el8_1
* kernel-2.4.0-1.el9_0

File: convert2rhel/checks.py
Comment (abadger) [inline convert2rhel/checks.py:371]: Thinking about the issues I raised earlier with different package versions being evaluated wrongly, I begin to think we should be using `rpm.labelCompare()` to do our sorting rather than trying to do it ourselves.  This is in part because we're not just scanning for kernel packages here, but also kmod packages (whose authors might be doing non-standard things with their naming or versioning).

The code for the two functions would look like this once ported to `rpm.labelCompare`:
``` python
import functools
import rpm

def get_most_recent_unique_kernel_pkgs(pkgs):
    pkgs_groups = itertools.groupby(sorted(pkgs), lambda pkg_name: pkg_name.split(":")[0])
    list_of_sorted_pkgs = []
    for distinct_kernel_pkgs in pkgs_groups:
        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod")):
            list_of_sorted_pkgs.append(max(distinct_kernel_pkgs[1], key=functools.cmp_to_key(_package_version_cmp)))

    return tuple(list_of_sorted_pkgs)

def _package_version_cmp(pkg_1, pkg_2):
    pkg_ver_components = []
    for pkg in pkg_1, pkg_2:
        # Remove the package name and split the rest between epoch + version and release + arch
        epoch_version, release_arch =  pkg.rsplit('-', 2)[-2:]
        # Separate the (optional) epoch from the version
        epoch_version = epoch_version.split(':', 1)
        if len(epoch_version) == 1:
            epoch = '0',
            verson = epoch_version[0]
        else:
            epoch, version = epoch_version
        # Discard the arch
        release = release_arch.rsplit('.', 1)[0]
        pkg_ver_components.append((epoch, version, release))

    return rpm.labelCompare(pkg_ver_components[0], pkg_ver_components[1])
```
The splitting could also be done via a regex.  I believe we have such a regex in pkghandler.py although we would want to audit before using it (for two reasons: (1) yum and repoquery commands output epoch in different ways depending on the command [0:foo-1.0-2.el8.x86_64 vs foo-0:1.0-2.el8.x86_64].  Sometimes an epoch of 0 is omitted and sometimes it is placed in the string, and (2) some of the code we've written cares about the arch while others do not so the rege may grab arch or it might not.

@r0x0d Do you think that switching to `rpm.labelCompare` should be in this PR or a different one?

**Comment by abadger** [comment]
I've reviewed this.  Other than the questions about `get_most_recent_unique_kernel_pkgs()` and its helper functions, this looks good.

File: convert2rhel/systeminfo.py
Comment (abadger) [reply convert2rhel/systeminfo.py:313]: I can see the potential problem with repos not being enabled yet but I'm not sure whether that is ever a problem in actuality (especially since customers aren't reporting failures because of it).  @zhukovgreen Do you happen to know about this TODO and whether we can either enhance what it says or remove it?

Comment (zhukovgreen) [reply convert2rhel/systeminfo.py:313]: I don't have any input on this. Seems like some artifact which I forgot to remove

File: convert2rhel/checks.py
Comment (r0x0d) [reply convert2rhel/checks.py:371]: @abadger, I think we can include this switch in this PR without a problem. Although this is a code cleanup and not a refactor, I feel that cleaning up code sometimes leads to a better way of doing things. 

I will take a closer look at the code you provided and see if I can implement it!

Comment (r0x0d) [reply convert2rhel/checks.py]: @abadger I have included your piece of code in the last commit.

File: convert2rhel/pkghandler.py
Comment (abadger) [inline convert2rhel/pkghandler.py:1087]: Hmmm... A  few higher level thoughts on this function:

* Sometimes yum and dnf return NEVR like this: `NetworkManager-1:1.18.8-2.0.1.el7_9.x86_64`.  Other times they return it like this (ENVR form): `1:NetworkManager-1.18.8-2.0.1.el7_9.x86_64`.  Should we handle both of these cases?  In one function or in two?

* There's already a `pkghandler.compare_package_versions` function.  That one takes its input as EVR strings: `1:1.18.8-2.0.1.el7_9.x86_64` which removes the ambiguity around whether the epoch is before or after name because there is no name.  However, we can't use that directly as a key function (since the key will be a NEVR or an ENVR string).

* Should we verify that the name is the same for both strings passed into this function?  If I call the function with `package_version_cmp("test-1.0-1", "foobar-2.0-1") I would expect it to error rather than say foobar is later than test.

* Should we parse arch off the end of the NEVR (arch is `x86_64`, `i686`, `s390x`, etc) and error if arches do not match?

Here's my strawman to address all of those:

* Separate version parsing and version comparison
* The parsing function will take a version string in either ENVR or NEVR format. 
  * If epoch or arch does not exist in the input string, it will be set to `None` in the tuple.
  * Parsing arch will be slightly tricky as I don't think there's a fool-proof way to separate an optional arch from release.  Maybe looking for a list of allowed arch patterns is the best we can do. Something like: `.*(x86_64|s390x|i.86)?$`
* The version comparison function will take two version strings in either ENVR or NEVR format.
* The strings will be passed to the parsing functions to return the information in a normalized form.
  * The names and arches will be compared. 
    * An error will be emitted if names do not match.
    * An error will be emitted if arches do not match and both arches are not None.
* rpm.labelCompare() will then be used to compare the EVR for both packages and the results returned.
* Port existing use of `pkghandler.compare_package_versions()` to use the new comparison function.
* Remove `pkghandler.compare_package_versions()` and `utils.string_to_version()`.

Discuss?

Comment (r0x0d) [reply convert2rhel/pkghandler.py:1087]: resolving this one since it's a duplicate.

Comment (r0x0d) [reply convert2rhel/pkghandler.py:1087]: @abadger, I think we can follow up on this in our 1:1 sessions. Probably next week, we can discuss it a bit more on which direction we want to move with this function.  

Definitely, there is a lot of room to improve the current function.

**Comment by kokesak** [comment]
@r0x0d once you will have a free time, please rebase the PR

File: convert2rhel/pkghandler.py
Comment (r0x0d) [reply convert2rhel/pkghandler.py:1093]: I will create a Jira issue as well for specifically this one change.

**Comment by r0x0d** [comment]
@kokesak rebased!

**Comment by r0x0d** [comment]
I will rebase this to resolve the conflict with the checks.py

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:301]:
```suggestion
                "Prevent the unsupported modules from loading by following https://access.redhat.com/solutions/41278 and run convert2rhel again to continue with the conversion."
```

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:290]: There's still one part of the RHELC-297 missing:

> Improve readability/maintainability of the `ensure_compatibility_of_kmods` function in checks.py as per https://github.com/oamg/convert2rhel/pull/193#discussion_r599881876

```python
def get_unsupported_kmods(host_kmods, rhel_supported_kmods):
    """Return a set of full paths to those installed kernel modules that are not available in RHEL repositories.
    ...
    """
    unsupported_kmods_subpaths = host_kmods - rhel_supported_kmods - set(system_info.kmods_to_ignore)
    unsupported_kmods_full_paths = ["/lib/modules/{kver}/{kmod}".format(kver=kernel_version, kmod=kmod)
                                    for kmod in unsupported_kmods_subpaths]
    return unsupported_kmods_full_paths
```

```python
def ensure_compatibility_of_kmods():
    ...
    if not unsupported_kmods:
        logger.debug("Kernel modules are compatible.")
    else:
        logger.critical(
            (
                "The following kernel modules are not supported in RHEL:\n{kmods}\n"
                "Make sure you have updated the kernel to the latest available version and rebooted the system.\n"
                "Prevent the unsupported modules from loading by following https://access.redhat.com/solutions/41278"
                " and run convert2rhel again to continue with the conversion."
            ).format(kmods="\n".join(unsupported_kmods), system=system_info.name)
        )
```

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (bocekm) [reply convert2rhel/checks.py:293]: Resolved by https://github.com/oamg/convert2rhel/pull/469/commits/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.

File: convert2rhel/checks.py
Comment (bocekm) [reply convert2rhel/checks.py]: Resolved by https://github.com/oamg/convert2rhel/commit/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.

**Review by bocekm — APPROVED** [review]

File: convert2rhel/utils.py
Comment (bocekm) [inline convert2rhel/utils.py:509]: @r0x0d, you can remove this function and the associated unit test. It's not used anywhere anymore.

**Review by bocekm — CHANGES_REQUESTED** [review]

**Review by kokesak — CHANGES_REQUESTED** [review]
@danmyway since we want to improve the tests description, could you please update the summary in `tests/integration/tier0/inhibit-if-kmods-is-not-supported/main.fmf` and maybe even in the pytest file as well if you think it's not good enough

**Comment by danmyway** [comment]
> @danmyway since we want to improve the tests description, could you please update the summary in `tests/integration/tier0/inhibit-if-kmods-is-not-supported/main.fmf` and maybe even in the pytest file as well if you think it's not good enough

Oh, the test got overwritten by rebasing again, I will push it back.

**Comment by danmyway** [comment]
The test is back in @kokesak

**Comment by r0x0d** [comment]
whoops, I think we have a rebase problem in this PR. I will try to fix it

File: convert2rhel/utils.py
Comment (r0x0d) [reply convert2rhel/utils.py]: done.

**Comment by r0x0d** [comment]
Okay, I rebased that and applied the two latest commits from @danmyway here. Hope everything works as expected.

**Review by bocekm — DISMISSED** [review]

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:88]: I'm confused. The conversion should be inhibited no?

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:89]: In the beginning of the file is the check for custom kernel module no? Anyway I would probably removed this section completely

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:98]:
```suggestion
        assert c2r.expect("Tainted kernel modules detected") == 0
```
So it's the same as below or whatever of these two is correct?

**Review by kokesak — CHANGES_REQUESTED** [review]
Just a few small tweaks in the description in comments, other than that it looks good @danmyway

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (bocekm) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:88]: A force loaded kernel module is considered tainted as well and the conversion is supposed to be inhibited as well.

Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:88]: Oh yeah, I have changed the function name and did not apply the changes to the docstring

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (bocekm) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:98]: I changed the log message in https://github.com/oamg/convert2rhel/commit/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f. @r0x0d, is there a reason why you dropped the commit?

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:89]: This was added to check for handling scenarios when module is force loaded. 
The custom one is just loaded from a different custom path, thus considered custom.

Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:89]: My bad, the kmod being force loaded should inhibit the conversion.

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (r0x0d) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: This PR had a rebase problem where we had a lot of commits from main here... I just pushed my local changes and applied Daniel's commits on top of it.

Comment (r0x0d) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: @bocekm do you mean that the commit is not available here anymore or the change proposed is missing?

Comment (bocekm) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: Both. The commit is not among the PR commits and the changes from the commit are not part of any other PR commit.

Comment (r0x0d) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: I see. Could you please comment again the change so I can apply it?

Comment (bocekm) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: All the changes are already in https://github.com/oamg/convert2rhel/commit/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.

You can use the approach @abadger suggested on Monday when you dropped @danmyway's commits from https://github.com/oamg/convert2rhel/pull/554:
```
$ git pull --rebase
$ wget https://github.com/oamg/convert2rhel/commit/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.patch
$ git am 5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.patch
```

Comment (r0x0d) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: Okay, I applied the changes from 5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f.

Comment (bocekm) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py]: Not completely, for some reason. The map/lambda block is still present.

**Review by kokesak — APPROVED** [review]
LGTM

File: convert2rhel/utils.py
Comment (r0x0d) [inline convert2rhel/utils.py:65]: @abadger, I took this code from python 3.10 functools module, should I do the same as you did here? https://github.com/oamg/convert2rhel/pull/540/files#diff-f82efabf7bb126100afd0472e3d8a5902da6576c6ddf75e3184cce9eb647c52aR58-R59

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:294]: The map/lambda code block is something I removed under the commit https://github.com/oamg/convert2rhel/commit/5f52da81b3e23fb6a03cb4e5c2f4d251cb6b700f. The commit has been dropped from the PR twice. Please apply the patch or let me know why is it better to keep the map/lambda then the list comprehension.

**Review by bocekm — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (r0x0d) [reply convert2rhel/checks.py:294]: You're correct, actually. The list comprehension is more readable. 

We were having some problems during the rebase/pull of this branch. I will make sure to update it on Monday.

Comment (r0x0d) [reply convert2rhel/checks.py]: done.

**Comment by r0x0d** [comment]
@danmyway I have rebased this PR to just one commit, hopefully, it will cause less confusion now.

**Review by bocekm — APPROVED** [review]
This looks good already from my POV! The tests fail though.

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:34]:
```suggestion
    This test verifies that rpmquery for detecting supported kernel modules in RHEL works correctly.
```

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:88]:
```suggestion
    Test force loads kmod and verifies that Convert2RHEL run is being inhibited.
```

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (kokesak) [inline tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:91]: @danmyway, If this is failing only on CentOS7/OL7 you can add a simple condition and skip this test case scenario on those el7 distros.

Unless you find a solution to make it work properly on EL7 distros.

**Comment by r0x0d** [comment]
@danmyway, since the conflict is in an integration tests, do you want to try to resolve yourself, or do you want me to take care of it?

File: convert2rhel/utils.py
Comment (abadger) [reply convert2rhel/utils.py:69]: @r0x0d Yes, that would be good. Keeping track of licenses and copyright when copying code is important.

File: convert2rhel/checks.py
Comment (bocekm) [inline convert2rhel/checks.py:297]:
```suggestion
            "First, make sure you have updated the kernel to the latest available version and rebooted the system.\n"
            "If this message appears again after doing the above, prevent the modules from loading by following {1}"
```
@abadger, @SpyTec, @r0x0d, as we discussed on the Cabal session, without having the kernel currency check available on OL 8, a message notifying the user that the kernel needs to be up-to-date is useful.
I was thinking whether to go into more detail, saying that the loaded kernel should be the latest available for the minor version of the system they are converting but:
1. the message would not be concise
2. it would be difficult to phrase in a way that works also for a distro without minor versions (Stream)

**Review by bocekm — CHANGES_REQUESTED** [review]

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:91]: Yep, that workaround would be needed for now.

**Review by bocekm — APPROVED** [review]

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:79]: Good catch, thank you :+1:

File: tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py
Comment (danmyway) [reply tests/integration/tier0/inhibit-if-kmods-is-not-supported/test_kmods_not_supported.py:89]: I'll fix that, thank you for spotting that!

File: convert2rhel/checks.py
Comment (SpyTec) [inline convert2rhel/checks.py:431]: Unit tests for centos6 says this doesn't exist, which is confirmed in 2.6 docs https://docs.python.org/2.6/library/functools.html

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: convert2rhel/checks.py
Comment (r0x0d) [reply convert2rhel/checks.py]: done.

**Review by kokesak — DISMISSED** [review]
The integration test looks nice.

File: convert2rhel/checks.py
Comment (r0x0d) [reply convert2rhel/checks.py]: done.

**Review by abadger — APPROVED** [review]
Code looks fine but there were two failing integration tests.  It looks like those are due to a problem with the yum repository.  I've kicked off new runs for both of those.

I also see that Eric had a crequest to change something but I wasn't able to find it in the converaation.  Can you check if that was addressed before merging?

**Review by SpyTec — APPROVED** [review]
> I also see that Eric had a crequest to change something but I wasn't able to find it in the converaation. Can you check if that was addressed before merging?

It was related to py2.6 and been addressed, LGTM as well

**Comment by r0x0d** [comment]
Thanks, @abadger for re-running the tests. Now the only thing that failed is the codecov patch/project checks, but they are fine. 

@kokesak could you approve it once again?

**Review by kokesak — APPROVED** [review]
Looks nice, merge it whenever you want.

**Pull request merged**
